File: src/types/project.ts
import { Task } from './task';

export enum ProjectStatus {
  PLANIFIE = "PLANIFIE",
  EN_COURS = "EN_COURS",
  EN_PAUSE = "EN_PAUSE",
  TERMINE = "TERMINE",
  ANNULE = "ANNULE"
}

export interface ProjectStats {
  projets_actifs: number;
  total_projets: number;
  variation_projets_actifs: number;
  taches_completees: number;
  total_taches: number;
  taches_retard: number;
  variation_taches_retard: number;
  heures_travaillees: number;
  variation_heures: number;
  repartition: {
    en_cours: number;
    en_attente: number;
    termines: number;
    en_retard: number;
  };
  taux_completion: number;
  projets_termines: number;
}

export interface ProjectBase {
  code: string;
  nom: string;
  description?: string;
  date_debut: string;
  date_fin_prevue: string;
  date_fin_reelle?: string;
  statut: ProjectStatus;
  budget?: number;
  responsable_id: string;
  objectifs?: Array<{
    description: string;
    criteres_succes?: string;
    date_cible?: string;
    statut?: string;
  }>;
  risques?: Array<{
    description: string;
    impact: 'FAIBLE' | 'MOYEN' | 'ELEVE';
    probabilite: 'FAIBLE' | 'MOYENNE' | 'ELEVEE';
    mitigation?: string;
    statut?: string;
  }>;
}

export interface Project extends ProjectBase {
  id: string;
  created_at: string;
  updated_at: string;
  taches?: Task[];
  documents?: ProjectDocument[];
}

export interface ProjectDocument {
  id: string;
  projet_id: string;
  nom: string;
  type: string;
  chemin_fichier: string;
  date_upload: string;
  uploaded_by_id: string;
}

export interface ProjectList {
  projects: Project[];
  total: number;
  page: number;
  size: number;
  total_pages: number;
}

File: models/project.py
import enum
import uuid
from sqlalchemy import Column, String, Float, Enum, JSON, ForeignKey, Text, Numeric, Date, DateTime
from sqlalchemy.orm import relationship
from .base import Base, BaseModel
from datetime import datetime
from typing import Optional
from sqlalchemy.dialects.postgresql import UUID


class ProjectStatus(enum.Enum):
    """Statuts possibles d'un projet"""
    PLANIFIE = "PLANIFIE"
    EN_COURS = "EN_COURS"
    EN_PAUSE = "EN_PAUSE"
    TERMINE = "TERMINE"
    ANNULE = "ANNULE"

class Project(Base):
    """Modèle représentant un projet"""
    __tablename__ = "projets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    code = Column(String(50), unique=True, index=True, nullable=False)
    nom = Column(String(200), nullable=False)
    description = Column(Text)
    date_debut = Column(Date, nullable=False)
    date_fin_prevue = Column(Date, nullable=False)
    date_fin_reelle = Column(Date)
    statut = Column(Enum(ProjectStatus), default=ProjectStatus.PLANIFIE)    
    budget = Column(Numeric(10, 2))
    responsable_id = Column(UUID(as_uuid=True), ForeignKey("employes.id"))
    objectifs = Column(JSON)  # Liste des objectifs du projet
    risques = Column(JSON)  # Liste des risques identifiés
    
    # Relations
    responsable = relationship("Employe", back_populates="projets_geres")
    taches = relationship("Tache", back_populates="projet")
    documents = relationship("ProjectDocument", back_populates="projet")

class ProjectDocument(BaseModel):
    """Modèle représentant un document lié à un projet"""
    __tablename__ = "documents_projet"

    projet_id = Column(UUID(as_uuid=True), ForeignKey("projets.id"), nullable=False)
    nom = Column(String(200), nullable=False)
    type = Column(String(50))  # Type de document (contrat, rapport, etc.)
    chemin_fichier = Column(String(500), nullable=False)  # Chemin vers le fichier
    date_upload = Column(DateTime, default=lambda: datetime.now(datetime.timezone.utc), nullable=False)
    uploaded_by_id = Column(UUID(as_uuid=True), ForeignKey("employes.id"))
    
    # Relations
    projet = relationship("Project", back_populates="documents")
    uploaded_by = relationship("Employe")
